   Aprendizado de máquina
   
   O aprendizado de máquina é uma área da Inteligência Artificial que busca o desenvolvimento de técnicas computacionais sobre o aprendizado e, também, a construção de sistemas que consigam adquirir conhecimento de maneira automática, sendo um sistema de aprendizado um programa de computador que toma decisões através de experiências acumuladas à partir de soluções bem sucedidas de problemas anteriores.(Monard & Baranauskas, 2003).
   Ou seja, aprendizagem de máquina estuda métodos computacionais para melhorar o desempenho, através da mecanização da aquisição de conhecimento baseado em experiências. (Langley & Simon, 1995)
    Dessa forma, podemos utilizar o aprendizado de máquina para solucionar diversos problemas, como por exemplo:
        -Processamento de Linguagem natural
        -Visão Computacional (processamento de imagens)
        -Análise de sequenciamento genético
        -Engenharia de Software
        -etc.


   Aprendizado Supervsionado e Não Supervisionado
    
   O aprendizado indutivo, que se efetua através de raciocínio sobre exemplos fornecidos por um processo externo ao sistema de aprendizado, pode ser dividido em dois subconjuntos, o supervisionado e o não-supervisionado. No aprendizado supervisionado, depende-se de um conjunto de exemplos de treinamento para serem fornecidos ao algoritmo de aprendizado, para que o mesmo possa realizar uma inferência posterior sobre novos exemplos.Já o não supervisionado, independe de um conjunto para treinamento, analisando os exemplos fornecidos, ele busca agrupá-los de alguma maneira, geralmente necessitando que, posteriormente, esses agrupamentos sejam analisados para definir o significado de cada agrupamento no contexto analisado. (Monard & Baranauskas, 2003)
   A partir disso, iremos abordar no curso dois algoritmos para cada tipo de aprendizado, realizando Clustering e Topic Modeling (Não-supervisionados). Já no supervisionado, utilizaremos Naive Bayes e Redes Neurais.


   Scikit Learn
    
   Scikit-Learn é um módulo em Python que integra uma vasta gama de algoritmos de aprendizado de máquina para problemas supervisionados ou não-supervisionados de média escala, buscando facilitar o uso de aprendizado de máquina para não especialistas, usando uma linguagem de alto nível para propósito geral.(Pedregosa et al., 2011).
   Sendo assim, o scikit será fundamental para os nossos exercícios, visto que ele cria uma camada de abstração para os algoritmos de aprendizagem de máquina, fazendo com que nos preocupemos no conceito dos algoritmos e sua aplicação, e não necessariamente à implementação dos mesmos.


   Classificador Naive Bayes
    
   O classificador Naive Bayes é um método designado para classificação de dados através do aprendizado indutivo supervisionado, que busca performar uma predição correta das classes de testes, possuindo uma abordagem simples e com conhecimento probabilístico. (John & Langley, 1995)
   Além disso, este classificador é um dos mais eficientes em processamento e precisão em relação à rotulação de novas amostras.(Schmitt, 2013).
   Não iremos aprofundar os conceitos deste modelo probabilístico, porém, recomendo vejam as seguintes video aulas do canal Data Science Academy, no youtube, para entender melhor como ele funciona, e, caso queira um detalhamento maior e mais científico, a leitura do artigo de John e Langley, que estará nas referências.
   
    Cap 12 - Como é Construído um Classificador Baseado no Teorema de Bayes? - Parte 1/2:	https://www.youtube.com/watch?v=JH8297s4D2g
    Cap 12 - Como é Construído um Classificador Baseado no Teorema de Bayes? - Parte 2/2:	https://www.youtube.com/watch?v=hvgs-L56jjM
    

   Implementação do código

   Visto que já temos um conhecimento básico acerca dos conceitos, algoritmos e ferramentas que vamos utilizar, começaremos agora a implementar um código que, através do algoritmo Gaussian Naive Bayes disponibilizado pelo Scikit Learn, consiga predizer se uma pessoa possui Diabetes ou não à partir de um conjunto de dados de saúde da pessoa.


   Primeiramente, vamos importar as bibliotecas necessárias

In [45]:
import pandas as pd #Necessária para manipular o dataset
import numpy as np
import matplotlib as plt
from sklearn.naive_bayes import GaussianNB #Importando o algoritmo Naive Bayes

   Para criarmos o nosso modelo de classificador  Naive Bayeisiano Gaussiano é bem simples, como podemos ver, com apenas uma linha conseguimos criar uma instância do modelo para posteriormente treiná-lo e utilizá-lo para realizar rotulações de novos dados.


In [46]:
#Criando uma instância do modelo Gaussian Naive Bayes e atribuindo ele
#à váriavel model
model = GaussianNB()

   Agora iremos abrir nosso conjunto de dados sobre diversas pessoas que possuem diabetes ou não. Para isso, iremos fazer como foi visto na aula sobre Pandas, abrindo o arquivo "diabetes.csv" e criando um dataframe chamado dados com este dataset.


In [47]:
dados = pd.read_csv("diabetes.csv")
dados

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

   Como se pode ver, temos 768 linhas, cada uma representando uma pessoa, e nove colunas, que representam as características que irão ser utilizadas na inferência para predizer se uma pessoa possui diabetes ou não. No nosso caso, a coluna Outcome está representando o dado que queremos predizer, sendo assim, caso seja 0 a pessoa não possui diabetes, caso seja 1 ela possui esta condição.


   Para podermos começar o treino do nosso modelo, precisamos separar os dados em nosso dataframe, separando a coluna que queremos predizer das demais. Para isso vamos realizar uma pequena manipulação dos nomes das colunas do nosso dataframe, como será visto a seguir.


In [48]:
#separa o nome do dado/coluna que vai ser predizido
coluna_alvo=['Outcome'] 
#criamos uma lista com os nomes das colunas/dados que são utilizados para
#a inferencia do dado que iremos predizer, ou seja, dados independentes.
preditores = list(set(list(dados.columns))-set(coluna_alvo)) 


   Como podemos ver, através das variáveis preditores e coluna_alvo, que são listas com os nomes das colunas, podemos acessar mais fácil e intuitivamente o dataframe para manipulá-lo mais a seguir.


In [49]:
dados[preditores]

Pregnancies  BloodPressure   BMI  Glucose  Insulin  Age  \
0              6             72  33.6      148        0   50   
1              1             66  26.6       85        0   31   
2              8             64  23.3      183        0   32   
3              1             66  28.1       89       94   21   
4              0             40  43.1      137      168   33   
..           ...            ...   ...      ...      ...  ...   
763           10             76  32.9      101      180   63   
764            2             70  36.8      122        0   27   
765            5             72  26.2      121      112   30   
766            1             60  30.1      126        0   47   
767            1             70  30.4       93        0   23   

     DiabetesPedigreeFunction  SkinThickness  
0                       0.627             35  
1                       0.351             29  
2                       0.672              0  
3                       0.167             23  
4                       2.288             35  
..                        ...            ...  
763                     0.171             48  
764                     0.340             27  
765                     0.245             23  
766                     0.349              0  
767                     0.315             31  

[768 rows x 8 columns]

In [50]:
dados[coluna_alvo]

Outcome
0          1
1          0
2          1
3          0
4          1
..       ...
763        0
764        0
765        0
766        1
767        0

[768 rows x 1 columns]

   A partir de agora, iremos pegar os valores contidos no dataframe para que possamos aplicar no modelo para treiná-lo. Iremos separar em X, todos os valores que serão necessários para realizar a predição da diabetes de alguma pessoas. Já em Y, será justamente os valores ( 0  ou 1) que definem se a pessoa tem diabetes ou não.


In [51]:
X= dados[preditores].values
X

array([[ 6.   , 72.   , 33.6  , ..., 50.   ,  0.627, 35.   ],
       [ 1.   , 66.   , 26.6  , ..., 31.   ,  0.351, 29.   ],
       [ 8.   , 64.   , 23.3  , ..., 32.   ,  0.672,  0.   ],
       ...,
       [ 5.   , 72.   , 26.2  , ..., 30.   ,  0.245, 23.   ],
       [ 1.   , 60.   , 30.1  , ..., 47.   ,  0.349,  0.   ],
       [ 1.   , 70.   , 30.4  , ..., 23.   ,  0.315, 31.   ]])

In [52]:
Y= dados[coluna_alvo].values
Y

array([[1],
       [0],
       [1],
       [0],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
    

   Como pode ser visto, geramos vetores com uma quantidade de elementos igual ao número de linhas do dataframe, sendo cada elemento do vetor, outro vetor, porém com os dados de cada coluna.


   Agora iremos importar um método disponível no SK Learn, que nos permite separar nossos dados em variáveis, para realizarmos o treino e teste do nosso modelo.Dessa forma, teremos um X_train e X_teste, que são subconjuntos de X visto anteriormente, e, de forma análoga, y_train e y_test.
    A função train_test_split possui alguns parâmetros, para ver todos, veja a documentação disponível no site do Scikit Learn. No nosso código, como podemos ver abaixo, estamos usando X (conjunto de variáveis independentes), Y (conjunto dos dados que queremos obter através das predições), test_size para configuração da divisão ( 0.3 implica que serão 70% de treino e 30% de teste)
    Por fim, o random_state funciona da seguinte forma:
        -Caso você não especifique um número, toda vez que você treinar o modelo, ele irá gerar um resultado diferente.
        -Caso você especifique, a "configuração" daquele treino realizado ficará atrelado à aquele número, ou seja, toda vez que você usar um número ele irá reproduzir a mesma configuração.


In [53]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=0.3, random_state=0)

   Agora, com os dados devidamente separados, utilizaremos o método fit() para treinarmos o modelo, como é visto abaixo:

In [54]:
#Utilizamos o método fit em nosso modelo "model", passando como parametro
#Os dados de treino
model.fit(X_train,y_train)

C:\Users\PC\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GaussianNB(priors=None, var_smoothing=1e-09)

   Com o modelo treinado, pode-se agora realizar predições sobre dados novos. Para isto, vamo utilizar o método predict(), que recebe como parâmetro os dados que queremos predizer. Sendo assim, ao utilizarmos esse método, ele irá nos retornar um conjunto de valores, no mesmo formato de y_train ou y_test, então iremos guardar em uma nova variável para realizarmos avaliações métricas.


In [55]:
#Aplicando o teste e guardando o resultado em y_pred
y_pred = model.predict(X_test)

   Pronto, com tudo isto feito, já temos nosso modelo treinado e testado, sendo assim, uma boa opção é realizar análises sobre os resultados obtidos. Uma simples métrica de eficiência seria ver a acurácia obtida pelo nosso modelo.Através do metrics podemos verificar isso facilmente, como se pode ver abaixo.


In [56]:
#Importando a biblioteca de análise métrica.
from sklearn import metrics
#Através do método accuracy_score(), podemos passar como parametro o nossos
#dados que foram separados para o teste e os valores que foram obtidos através
#da predição do modelo
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.7619047619047619


   Caso se deseje uma visão mais detalhada dos resultados obtidos, existe um outro método, também contida em metrics, que nos gera uma tabela com outros escores de eficiência, como precision, recall e f1-score. Também podemos montar uma matriz de confusão, Vejamos abaixo:


In [57]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[138  19]
 [ 36  38]]
              precision    recall  f1-score   support

           0       0.79      0.88      0.83       157
           1       0.67      0.51      0.58        74

    accuracy                           0.76       231
   macro avg       0.73      0.70      0.71       231
weighted avg       0.75      0.76      0.75       231



Referencias

JOHN, George H.; LANGLEY, Pat. Estimating continuous distributions in Bayesian classifiers. arXiv preprint arXiv:1302.4964, 2013.

MONARD, Maria Carolina; BARANAUSKAS, José Augusto. Conceitos sobre aprendizado de máquina. Sistemas inteligentes-Fundamentos e aplicações, v. 1, n. 1, p. 32, 2003.

SCHMITT, Vinícius Fernandes. Uma análise comparativa de técnicas de aprendizagem de máquina para prever a popularidade de postagens no facebook. 2013.

LANGLEY, Pat; SIMON, Herbert A. Applications of machine learning and rule induction. Communications of the ACM, v. 38, n. 11, p. 54-64, 1995.

Scikit-learn: Machine Learning in Python, Pedregosa et al., JMLR 12, pp. 2825-2830, 2011.